pkg/server/v3/delta_test.go

```diff
@@ -0,0 +1,393 @@
+
+				// We should only have 7 watch channels initialized since that is the base map length
+				if config.deltaCounts[typ] != 1 {
+					t.Errorf("watch counts for typ: %s => got %v, want 1", typ, config.deltaCounts[typ])
+				}
+				if v := res.GetSystemVersionInfo(); v != "" {
+					t.Errorf("expected emtpy version on initial request, got %s", v)
+				}
+			case <-time.After(1 * time.Second):
+				t.Fatalf("got no response")
+			}
+		})
+	}
+}
+
+func TestDeltaWatchClosed(t *testing.T) {
+	for _, typ := range testTypes {
+		t.Run(typ, func(t *testing.T) {
+			config := makeMockConfigWatcher()
+			config.closeWatch = true
+			s := server.NewServer(context.Background(), config, server.CallbackFuncs{})
+
+			resp := makeMockDeltaStream(t)
+			resp.recv <- &discovery.DeltaDiscoveryRequest{
+				Node:    node,
+				TypeUrl: typ,
+			}
+
+			// Verify that the response fails when the watch is closed
+			if err := s.DeltaAggregatedResources(resp); err == nil {
+				t.Error("DeltaAggregatedResources() => got no error, want watch failed")
+			}
+
+			close(resp.recv)
+		})
+	}
+}
+
+func TestSendDeltaError(t *testing.T) {
+	for _, typ := range testTypes {
+		t.Run(typ, func(t *testing.T) {
+			config := makeMockConfigWatcher()
+			config.deltaResponses = makeDeltaResponses()
+			s := server.NewServer(context.Background(), config, server.CallbackFuncs{})
+
+			// make a request with an error
+			resp := makeMockDeltaStream(t)
+			resp.sendError = true
+			resp.recv <- &discovery.DeltaDiscoveryRequest{
+				Node:    node,
+				TypeUrl: typ,
+			}
+
+			// check that response fails since we expect an error to come through
+			if err := s.DeltaAggregatedResources(resp); err == nil {
+				t.Error("DeltaAggregatedResources() => got no error, want send error")
+			}
+
+			close(resp.recv)
+		})
+	}
+}
+
+func TestDeltaAggregatedHandlers(t *testing.T) {
+	config := makeMockConfigWatcher()
+	config.deltaResponses = makeDeltaResponses()
+	resp := makeMockDeltaStream(t)
+
+	resp.recv <- &discovery.DeltaDiscoveryRequest{
+		Node:    node,
+		TypeUrl: rsrc.ListenerType,
+	}
+	resp.recv <- &discovery.DeltaDiscoveryRequest{
+		Node:    node,
+		TypeUrl: rsrc.ClusterType,
+	}
+	resp.recv <- &discovery.DeltaDiscoveryRequest{
+		Node:                   node,
+		TypeUrl:                rsrc.EndpointType,
+		ResourceNamesSubscribe: []string{clusterName},
+	}
+	resp.recv <- &discovery.DeltaDiscoveryRequest{
+		TypeUrl:                rsrc.RouteType,
+		ResourceNamesSubscribe: []string{routeName},
+	}
+	resp.recv <- &discovery.DeltaDiscoveryRequest{
+		TypeUrl:                rsrc.SecretType,
+		ResourceNamesSubscribe: []string{secretName},
+	}
+
+	s := server.NewServer(context.Background(), config, server.CallbackFuncs{})
+	go func() {
+		if err := s.DeltaAggregatedResources(resp); err != nil {
+			t.Errorf("DeltaAggregatedResources() => got %v, want no error", err)
+		}
+	}()
+
+	count := 0
+	for {
+		select {
+		case <-resp.sent:
+			count++
+			if count >= 5 {
```

Comment:
Comment on the significance of 5? Or could we put the above requests in a list so we can just do len of that?

Reply:
I like the request list idea :)